core/src/test/java/org/elasticsearch/recovery/RecoveriesCollectionTests.java

```diff
@@ -57,83 +64,112 @@ public void onRecoveryFailure(RecoveryState state, RecoveryFailedException e, bo
+                final int referencesToStore = recovery.status().store().refCount();
+                collection.resetRecovery(recoveryId, recovery.status().shardId());
+                try (RecoveriesCollection.RecoveryRef resetRecovery = collection.getRecovery(recoveryId)) {
+                    assertNotSame(recovery.status(), resetRecovery);
+                    assertSame(recovery.status().CancellableThreads(), resetRecovery.status().CancellableThreads());
+                    assertSame(recovery.status().indexShard(), resetRecovery.status().indexShard());
+                    assertSame(recovery.status().store(), resetRecovery.status().store());
+                    assertEquals(referencesToStore + 1, resetRecovery.status().store().refCount());
+                    assertEquals(currentAsTarget+1, shard.recoveryStats().currentAsTarget()); // we blink for a short moment...
+                    recovery.close();
+                    expectThrows(ElasticsearchException.class, () -> recovery.status().store());
+                    assertEquals(referencesToStore, resetRecovery.status().store().refCount());
+                }
+                assertEquals(currentAsTarget, shard.recoveryStats().currentAsTarget());
+            }
+            try (RecoveriesCollection.RecoveryRef resetRecovery = collection.getRecovery(recoveryId)) {
+                shards.recoverReplica(shard, (s, n) -> {
+                    assertSame(s, resetRecovery.status().indexShard());
+                    return resetRecovery.status();
+                }, false);
+            }
+            shards.assertAllEqual(numDocs);
+            assertNull("recovery is done", collection.getRecovery(recoveryId));
+        }
     }
 
-
-    long startRecovery(RecoveriesCollection collection) {
-        return startRecovery(collection, listener, TimeValue.timeValueMinutes(60));
+    long startRecovery(RecoveriesCollection collection, DiscoveryNode sourceNode, IndexShard shard) {
+        return startRecovery(collection,sourceNode, shard, listener, TimeValue.timeValueMinutes(60));
     }
 
-    long startRecovery(RecoveriesCollection collection, RecoveryTargetService.RecoveryListener listener, TimeValue timeValue) {
-        IndicesService indexServices = getInstanceFromNode(IndicesService.class);
-        IndexShard indexShard = indexServices.indexServiceSafe(resolveIndex("test")).getShardOrNull(0);
-        final DiscoveryNode sourceNode = new DiscoveryNode("id", LocalTransportAddress.buildUnique(), emptyMap(), emptySet(),
-            Version.CURRENT);
+    long startRecovery(RecoveriesCollection collection, DiscoveryNode sourceNode, IndexShard indexShard,
+                       RecoveryTargetService.RecoveryListener listener, TimeValue timeValue) {
+        final DiscoveryNode rNode = getDiscoveryNode(indexShard.routingEntry().currentNodeId());
+        indexShard.markAsRecovering("remote", new RecoveryState(indexShard.shardId(), false, RecoveryState.Type.REPLICA, sourceNode,
```

Comment:
I spent some time thinking how we can fold this into the ReplicationGroup in a clean way. I couldn't come up with anything fantastic - the best way I had was to make the actually recovery code in ReplicationGroup be a method on it's own which these tests can override and only do the recovery collection part. not sure how much cleaner that will be.